File: sdk/csharp/src/Providers/Nimail.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// nimail 渠道（www.nimail.cn）。POST 表单 API，无需认证，token 即邮箱地址本身。
/// 创建：POST /api/applymail (mail=前缀@nimail.cn)；收信：POST /api/getmails (mail=&amp;time=0)。
/// </summary>
public static class Nimail
{
    private const string Base = "https://www.nimail.cn";
    private static readonly Random Rand = new();

    private static string Ua()
    {
        var cfg = Config.Get();
        if (cfg.Headers is not null && cfg.Headers.TryGetValue("User-Agent", out var ua) && !string.IsNullOrEmpty(ua))
            return ua;
        return "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36";
    }

    private static Dictionary<string, string> FormHeaders() => new()
    {
        ["User-Agent"] = Ua(),
        ["Content-Type"] = "application/x-www-form-urlencoded",
        ["Origin"] = Base,
        ["Referer"] = $"{Base}/",
    };

    private static string RandomLocal(int n)
    {
        const string chars = "abcdefghijklmnopqrstuvwxyz0123456789";
        var sb = new StringBuilder();
        for (var i = 0; i < n; i++) sb.Append(chars[Rand.Next(chars.Length)]);
        return sb.ToString();
    }

    private static string First(JsonObject msg, params string[] keys)
    {
        foreach (var k in keys)
        {
            var v = Json.Str(msg, k);
            if (!string.IsNullOrEmpty(v)) return v;
        }
        return "";
    }

    /// <summary>创建 nimail.cn 临时邮箱</summary>
    public static EmailInfo Generate()
    {
        var email = $"{RandomLocal(10)}@nimail.cn";
        var body = $"mail={WebUtility.UrlEncode(email)}";
        var resp = Http.Post($"{Base}/api/applymail", body, "application/x-www-form-urlencoded", FormHeaders());
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject;
        if (data is null || Json.Str(data, "success") != "true" || string.IsNullOrEmpty(Json.Str(data, "user")))
            throw new Exception("nimail: 创建邮箱失败");
        var user = Json.Str(data, "user");
        return new EmailInfo("nimail", user, user);
    }

    /// <summary>获取 nimail.cn 邮件列表</summary>
    public static List<Email> GetEmails(string email)
    {
        var addr = (email ?? "").Trim();
        if (addr.Length == 0) throw new Exception("nimail: 缺少邮箱地址");
        var body = $"mail={WebUtility.UrlEncode(addr)}&time=0";
        var resp = Http.Post($"{Base}/api/getmails", body, "application/x-www-form-urlencoded", FormHeaders());
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject;
        var result = new List<Email>();
        if (data is null || Json.Str(data, "success") != "true") return result;
        if (data["mail"] is not JsonArray mail) return result;
        foreach (var item in mail)
        {
            if (item is not JsonObject msg) continue;
            var raw = new Dictionary<string, object?>
            {
                ["id"] = First(msg, "id", "time"),
                ["from"] = First(msg, "from", "sender"),
                ["to"] = addr,
                ["subject"] = First(msg, "subject", "title"),
                ["text"] = First(msg, "text", "content"),
                ["html"] = First(msg, "html", "content"),
                ["date"] = First(msg, "date", "time"),
            };
            result.Add(Normalize.NormalizeEmail(raw, addr));
        }
        return result;
    }
}

File: sdk/php/src/Providers/Nimail.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * nimail 渠道实现 — https://www.nimail.cn
 *
 * 简单 POST 表单 API，无需认证，Token 即邮箱地址本身。
 * 1. POST /api/applymail (mail=前缀@nimail.cn) → 注册邮箱
 * 2. POST /api/getmails (mail=&time=0) → 返回 {"mail": [...], "success": "true"}
 */
final class Nimail
{
    private const CHANNEL = 'nimail';
    private const BASE = 'https://www.nimail.cn';
    private const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36';

    /**
     * @return array<string,string>
     */
    private static function formHeaders(): array
    {
        return [
            'User-Agent' => self::UA,
            'Content-Type' => 'application/x-www-form-urlencoded',
            'Origin' => self::BASE,
            'Referer' => self::BASE . '/',
        ];
    }

    private static function randomLocal(int $n = 10): string
    {
        $chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
        $out = '';
        for ($i = 0; $i < $n; $i++) {
            $out .= $chars[random_int(0, strlen($chars) - 1)];
        }
        return $out;
    }

    /**
     * 按优先级从消息中提取首个非空值并转为字符串。
     *
     * @param array<mixed> $msg
     */
    private static function first(array $msg, string ...$keys): string
    {
        foreach ($keys as $key) {
            $val = $msg[$key] ?? null;
            if ($val !== null && $val !== '') {
                return (string) $val;
            }
        }
        return '';
    }

    public static function generate(): EmailInfo
    {
        $email = self::randomLocal() . '@nimail.cn';
        $resp = HttpClient::post(
            self::BASE . '/api/applymail',
            self::formHeaders(),
            body: 'mail=' . rawurlencode($email),
        );
        $data = HttpClient::json($resp);
        if (($data['success'] ?? null) !== 'true' || empty($data['user'])) {
            throw new \RuntimeException('nimail: 创建邮箱失败');
        }
        $user = (string) $data['user'];
        return new EmailInfo(self::CHANNEL, $user, $user);
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email, ?string $token = null): array
    {
        $addr = trim($email);
        if ($addr === '') {
            throw new \InvalidArgumentException('nimail: 缺少邮箱地址');
        }
        $resp = HttpClient::post(
            self::BASE . '/api/getmails',
            self::formHeaders(),
            body: 'mail=' . rawurlencode($addr) . '&time=0',
        );
        $data = HttpClient::json($resp);
        if (($data['success'] ?? null) !== 'true') {
            return [];
        }
        $mail = $data['mail'] ?? null;
        if (!is_array($mail)) {
            return [];
        }
        $out = [];
        foreach ($mail as $msg) {
            if (!is_array($msg)) {
                continue;
            }
            $row = [
                'id' => self::first($msg, 'id', 'time'),
                'from' => self::first($msg, 'from', 'sender'),
                'to' => $addr,
                'subject' => self::first($msg, 'subject', 'title'),
                'text' => self::first($msg, 'text', 'content'),
                'html' => self::first($msg, 'html', 'content'),
                'date' => self::first($msg, 'date', 'time'),
            ];
            $out[] = Normalize::email($row, $addr);
        }
        return $out;
    }
}
